File: frontend/zero_world/lib/models/listing.dart
class Listing {
  const Listing({
    required this.id,
    required this.title,
    required this.description,
    required this.price,
    required this.ownerId,
    required this.isActive,
    required this.createdAt,
    this.updatedAt,
    this.category,
    this.location,
    this.imageUrls = const [],
  });

  factory Listing.fromJson(Map<String, dynamic> json) {
    return Listing(
      id: json['_id'] as String,
      title: json['title'] as String? ?? '',
      description: json['description'] as String? ?? '',
      price: (json['price'] as num?)?.toDouble() ?? 0,
      ownerId: json['owner_id'] as String? ?? '',
      isActive: json['is_active'] as bool? ?? true,
      createdAt:
          DateTime.tryParse(json['created_at']?.toString() ?? '') ??
          DateTime.now(),
      updatedAt: json['updated_at'] != null
          ? DateTime.tryParse(json['updated_at']!.toString())
          : null,
      category: json['category'] as String?,
      location: json['location'] as String?,
      imageUrls:
          (json['image_urls'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          const [],
    );
  }

  Map<String, dynamic> toJson() => {
    '_id': id,
    'title': title,
    'description': description,
    'price': price,
    'owner_id': ownerId,
    'is_active': isActive,
    'created_at': createdAt.toIso8601String(),
    'updated_at': updatedAt?.toIso8601String(),
    'category': category,
    'location': location,
    'image_urls': imageUrls,
  };

  final String id;
  final String title;
  final String description;
  final double price;
  final String ownerId;
  final bool isActive;
  final DateTime createdAt;
  final DateTime? updatedAt;
  final String? category;
  final String? location;
  final List<String> imageUrls;
}

File: backend/app/schemas/listing.py
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field, ConfigDict


class ListingBase(BaseModel):
    title: str = Field(..., min_length=3, max_length=140)
    description: str = Field(..., min_length=10, max_length=4000)
    price: float = Field(..., ge=0)
    category: Optional[str] = Field(default=None, max_length=80)
    location: Optional[str] = Field(default=None, max_length=120)
    image_urls: List[str] = Field(default_factory=list)


class ListingCreate(ListingBase):
    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "title": "Mountain Bike",
                "description": "Lightly used trail bike in great condition.",
                "price": 450.0,
                "category": "Sports",
                "location": "Cape Town",
                "image_urls": [
                    "https://example.com/bike-front.jpg",
                    "https://example.com/bike-side.jpg",
                ],
            }
        }
    )


class ListingUpdate(BaseModel):
    title: Optional[str] = Field(default=None, min_length=3, max_length=140)
    description: Optional[str] = Field(default=None, min_length=10, max_length=4000)
    price: Optional[float] = Field(default=None, ge=0)
    category: Optional[str] = Field(default=None, max_length=80)
    location: Optional[str] = Field(default=None, max_length=120)
    image_urls: Optional[List[str]] = None
    is_active: Optional[bool] = None


class ListingPublic(ListingBase):
    id: str = Field(alias="_id")
    owner_id: str
    is_active: bool = True
    created_at: datetime
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(populate_by_name=True)
